packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+  }
+
+  *[Symbol.iterator]() {
+    yield* filter(this.iterable, this.filterFn);
+  }
+
+  get size() {
+    return this.keyMap.size;
+  }
+
+  getKeys() {
+    return this.keyMap.keys();
+  }
+
+  getKeyBefore(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.prevKey : null;
+  }
+
+  getKeyAfter(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.nextKey : null;
+  }
+
+  getFirstKey() {
+    return this.firstKey;
+  }
+
+  getLastKey() {
+    return this.lastKey;
+  }
+
+  getItem(key: Key) {
+    return this.keyMap.get(key);
+  }
+}
+
+let whitespace = /\s/;
+
+export function useComboBoxState<T extends object>(props: ComboBoxProps<T>): ComboBoxState<T> {
+  let {
+    onFilter,
+    selectedKey,
+    defaultSelectedKey,
+    inputValue,
+    defaultInputValue,
+    onInputChange,
+    isFocused,
+    isOpen,
+    defaultOpen
+  } = props;
+  
+  let itemsControlled = !!onFilter;
+  let collator = useCollator({sensitivity: 'base'});
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = inputValue !== undefined;
+  let selectedControlled = !!selectedKey;
+   */
+
+  // Create a separate menu open state tracker so onOpenChange doesn't fire with open and close in quick succession
+  // in cases where there aren't items to show
+  // Note that this means onOpenChange won't fire for controlled open states
+  let [menuIsOpen, setMenuIsOpen] = useControlledState(isOpen, defaultOpen || false, () => {});
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
```

Comment:
I don't think I'd compare character by character, because it may not work well on unicode strings. In some cases, a "character" is actually multiple surrogate pairs (what JS calls a character), e.g. emoji are sometimes between 2-6 JS characters. Instead, you could still loop by character, but slice a substring of the same length as the input string, and compare that with the collator all at once. If it doesn't match, then move the window forward by one character and try again.